### ISSUE TITLE
Handling spot termination signal for quickest replacement (stop re-adding to ASG)

### ISSUE THREAD

**Issue by phils** [cristim/autospotting#87]
## Issue type ##

- Documentation Report / Feature

## Build number #

```text
build261
```

## Environment ##

- ap-southeast-2
- VPC

## Summary ##

I'm trying to work out the best way to replace a spot instance that is about to terminate (to ideally avoid or minimise the time where zero instances are working in the AZ).

My idea is to check for the spot termination signal in metadata, then do the following on the instance when it hits:

- Detach itself from the ASG (so ASG will begin replacing it ASAP)
- Set timer to halt the instance after 3 minutes (just in case AWS doesn't kill it as promised)

However, during testing I found that autospotting just re-adds the spot instance into the ASG again and terminates the new on-demand one (all depends on the timing).

So now I'm thinking:

- Detach itself from the ASG (so ASG will begin replacing it ASAP)
- Determine the spot request ID, and remove the tag from it (to prevent autospotting from re-adding this instance to the ASG)
- Set timer to halt the instance after 3 minutes (just in case AWS doesn't kill it as promised)

I wonder if autospotting should use tagging so it avoids re-adding an instance to an ASG when it is removed?

Thanks.

**Comment by cristim** [comment]
Thanks for reporting this, I've been aware of issues with abrupt termination and recently started to work on it as well.

For the spot termination you can now use the [seespot] tool, which can run a program when the metadata is changed to indicate a termination.

I recently started writing a [program](https://github.com/cristim/instance-drainer) dedicated for this and I plan to enhance it going forward to entirely replace seespot. ASG detachment and tag removal from the spot specification and controlled termination could all be implemented there with little effort.

Autospotting just used the resource tags for a lot of its discovery, so once the tag was deleted from the spot request it would just ignore that instance

**Comment by phils** [comment]
We could perhaps avoid having to untag from the instance if autospotting itself kept track of which instances it had already added to the ASG.

If we agree that autospotting should never add an individual instance to the ASG more than once.. upon adding to ASG it could set an extra tag to flag this.  Then if it comes across it again, it can see this flag (or it could be filtered out on the initial query) so that it won't add it a second time.  Or alternatively, it could add a tag on initial spinup, that is removed once it is added to the ASG for the first time.

I think it's probably more desirable to add complexity into autospotting than having to implement it on every instance.

**Comment by cristim** [comment]
Sure, we can mark the spot request as processed by setting a tag on it from AutoSpotting as soon as the instance was added to the group, and next time we look at it we just ignore it if that tag is set so we only process the instance once.

We will need to run some stuff on the instance for intercepting the termination signal and trigger the detachment from the group, load balancer and ECS. The detach itself could be done from either AutoSpotting(which would need to be triggered via SNS by the instance) or by the instance itself, but wherever we do it we would require some IAM permissions on the instance and maybe also on AutoSpotting.

Doing it from AutoSpotting we can control the IAM permissions at deployment (or update) time ans we can extend the permissions gradually depending on which services we implemented detachment for, but we still need some IAM permission on the instance in order to use SNS.

Additionally the instance could set certain tags on itself to instruct AutoSpotting what it needs to do with it.  

The instance-drainer could be used either way, since it will be made portable.
